src/backend/optimizer/path/allpaths.c

```diff
@@ -2125,15 +2125,33 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 			subplan = subquery_planner(cteroot->glob, subquery, cteroot, cte->cterecursive,
 									   tuple_fraction, &subroot, config);
 
-			cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
+			if (!CdbPathLocus_IsSegmentGeneral(*subplan->flow))
+				cteplaninfo->subplan = prepare_plan_for_sharing(cteroot, subplan);
+			else
+				cteplaninfo->subplan = subplan;
+
 			cteplaninfo->subroot = subroot;
 		}
 
 		/*
 		 * Create another ShareInputScan to reference the already-created
-		 * subplan.
+		 * subplan if not avoiding sharing. Sharing SegmentGeneral subplan may
+		 * lead to deadlock when executed with 1-gang and joined with n-gang.
 		 */
-		subplan = share_prepared_plan(cteroot, cteplaninfo->shared_plan);
+		if (!CdbPathLocus_IsSegmentGeneral(*cteplaninfo->subplan->flow))
+		{
```

Comment:
Can we omit the parentheses here and below?
And also put comments from [lines](https://github.com/arenadata/gpdb/pull/640/files#diff-4eff9e11378a8d7655a2cd15eea18c2fa93687163c15bcdd6e4dd54818fef85fR2148-R2150) after [lines](https://github.com/arenadata/gpdb/pull/640/files#diff-4eff9e11378a8d7655a2cd15eea18c2fa93687163c15bcdd6e4dd54818fef85fR2137-R2139)?

Reply:
1) No, this is going to be a multi-line condition
2) Makes no sense to put the comment for the sub-if to the parent if